ipn/store/stores.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package store provides various implementation of ipn.StateStore.
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+
+	"tailscale.com/atomicfile"
+	"tailscale.com/ipn"
+	"tailscale.com/ipn/store/mem"
+	"tailscale.com/paths"
+	"tailscale.com/types/logger"
+)
+
+// Provider returns a StateStore for the provided path.
+type Provider func(path string) (ipn.StateStore, error)
+
+var regOnce sync.Once
+
+var registerAvailableExternalStores func()
+
+func registerDefaultStores() {
+	Register("mem:", mem.New)
+
+	if registerAvailableExternalStores != nil {
+		registerAvailableExternalStores()
+	}
+}
+
+var knownStores map[string]Provider
+
+// New returns a StateStore based on the provided path
+// and registered stores.
+// By default the following stores are registered:
+//
+//   * if the string begins with "mem:", the suffix
+//     is ignored and an in-memory store is used.
+//   * (Linux-only) if the string begins with "arn:",
+//     the suffix an AWS ARN for an SSM.
+//   * (Linux-only) if the string begins with "kube:",
+//     the suffix is a Kubernetes secret name
+//   * In all other cases, the path is treated as a filepath.
+func New(logf logger.Logf, path string) (ipn.StateStore, error) {
+	regOnce.Do(registerDefaultStores)
+	for prefix, sf := range knownStores {
+		if strings.HasPrefix(path, prefix) {
+			// We can't strip the prefix here as some NewStoreFunc (like arn:)
+			// expect the prefix.
+			return sf(path)
```

Comment:
should we pass the logger.Logf down to the Provider?

Reply:
Good idea, done.